File: Examples/OSX/AudioKitDemo/AudioKitDemo/Analysis/AnalysisViewController.m
//
//  AnalysisViewController.m
//  AudioKitDemo
//
//  Created by Aurelius Prochazka on 2/23/15.
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//

#import "AnalysisViewController.h"
#import "AKFoundation.h"

@implementation AnalysisViewController
{
    AKMicrophone *microphone;
    AKAudioAnalyzer *analyzer;

    IBOutlet NSTextField *frequencyLabel;
    IBOutlet NSTextField *amplitudeLabel;
    IBOutlet NSTextField *noteNameWithSharpsLabel;
    IBOutlet NSTextField *noteNameWithFlatsLabel;

    NSArray *noteFrequencies;
    NSArray *noteNamesWithSharps;
    NSArray *noteNamesWithFlats;

    AKSequence *analysisSequence;
    AKEvent *updateAnalysis;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    noteFrequencies = @[@16.35,@17.32,@18.35,@19.45,@20.6,@21.83,@23.12,@24.5,@25.96,@27.5,@29.14,@30.87];
    noteNamesWithSharps = @[@"C", @"C♯",@"D",@"D♯",@"E",@"F",@"F♯",@"G",@"G♯",@"A",@"A♯",@"B"];
    noteNamesWithFlats  = @[@"C", @"D♭",@"D",@"E♭",@"E",@"F",@"G♭",@"G",@"A♭",@"A",@"B♭",@"B"];

    AKSettings.shared.audioInputEnabled = YES;

    microphone = [[AKMicrophone alloc] init];
    [AKOrchestra addInstrument:microphone];
    analyzer = [[AKAudioAnalyzer alloc] initWithInput:microphone.output];
    [AKOrchestra addInstrument:analyzer];
}

- (void)viewDidAppear
{
    [super viewDidAppear];
    [analyzer start];
    [microphone start];

    analysisSequence = [AKSequence sequence];
    updateAnalysis = [[AKEvent alloc] initWithBlock:^{
        [self updateUI];
        [analysisSequence addEvent:updateAnalysis afterDuration:0.1];
    }];
    [analysisSequence addEvent:updateAnalysis];
    [analysisSequence play];
}

- (void)viewWillDisappear {
    [analyzer stop];
    [microphone stop];
}

- (void)updateUI {

    if (analyzer.trackedAmplitude.value > 0.1) {
        frequencyLabel.stringValue = [NSString stringWithFormat:@"%0.1f", analyzer.trackedFrequency.value];

        float frequency = analyzer.trackedFrequency.value;
        while (frequency > [noteFrequencies.lastObject floatValue]) {
            frequency = frequency / 2.0;
        }
        while (frequency < [noteFrequencies.firstObject floatValue]) {
            frequency = frequency * 2.0;
        }

        float minDistance = 10000;
        int index =  0;
        for (int i = 0; i < noteFrequencies.count; i++) {
            float distance = fabs([noteFrequencies[i] floatValue] - frequency);
            if (distance < minDistance) {
                index = i;
                minDistance = distance;
            }
        }
        int octave = (int)log2f(analyzer.trackedFrequency.value / frequency);
        NSString *noteName = [NSString stringWithFormat:@"%@%d", noteNamesWithSharps[index], octave];
        noteNameWithSharpsLabel.stringValue = noteName;
        noteName = [NSString stringWithFormat:@"%@%d", noteNamesWithFlats[index], octave];
        noteNameWithFlatsLabel.stringValue = noteName;

        [frequencyLabel setNeedsDisplay];
        [amplitudeLabel setNeedsDisplay];
        [noteNameWithSharpsLabel setNeedsDisplay];
        [noteNameWithFlatsLabel setNeedsDisplay];
    }
    amplitudeLabel.stringValue = [NSString stringWithFormat:@"%0.2f", analyzer.trackedAmplitude.value];

}
@end

File: Examples/OSX/Swift/AudioKitDemo/AudioKitDemo/AnalysisViewController.swift
//
//  AnalysisViewController.swift
//  AudioKitDemo
//
//  Created by Nicholas Arner on 3/1/15.
//  Copyright (c) 2015 AudioKit. All rights reserved.
//

class AnalysisViewController: NSViewController {

    @IBOutlet var frequencyLabel: NSTextField!
    @IBOutlet var amplitudeLabel: NSTextField!
    @IBOutlet var noteNameWithSharpsLabel: NSTextField!
    @IBOutlet var noteNameWithFlatsLabel: NSTextField!

    var analyzer: AKAudioAnalyzer!
    let microphone = AKMicrophone()

    let noteFrequencies = [16.35,17.32,18.35,19.45,20.6,21.83,23.12,24.5,25.96,27.5,29.14,30.87]
    let noteNamesWithSharps = ["C", "C♯","D","D♯","E","F","F♯","G","G♯","A","A♯","B"]
    let noteNamesWithFlats = ["C", "D♭","D","E♭","E","F","G♭","G","A♭","A","B♭","B"]

    let analysisSequence = AKSequence()
    var updateAnalysis: AKEvent?

    override func viewDidLoad() {
        super.viewDidLoad()

        AKSettings.shared().audioInputEnabled = true

        analyzer = AKAudioAnalyzer(input: microphone.output)

        AKOrchestra.addInstrument(microphone)
        AKOrchestra.addInstrument(analyzer)
        analyzer.play()
        microphone.play()

        let analysisSequence = AKSequence()
        updateAnalysis = AKEvent {
            self.updateUI()
            analysisSequence.addEvent(self.updateAnalysis, afterDuration: 0.1)
        }
        analysisSequence.addEvent(updateAnalysis)
        analysisSequence.play()
    }

    func updateUI() {
        if (analyzer.trackedAmplitude.floatValue > 0.1) {
            frequencyLabel.stringValue = String(format: "%0.1f", analyzer.trackedFrequency.floatValue)

            var frequency = analyzer.trackedFrequency.floatValue
            while (frequency > Float(noteFrequencies[noteFrequencies.count-1])) {
                frequency = frequency / 2.0
            }
            while (frequency < Float(noteFrequencies[0])) {
                frequency = frequency * 2.0
            }
            var minDistance: Float = 10000.0
            var index = 0

            for (var i = 0; i < noteFrequencies.count; i++) {

                let distance = fabsf(Float(noteFrequencies[i]) - frequency)
                if (distance < minDistance){
                    index = i
                    minDistance = distance
                }
            }

            let octave = Int(log2f(analyzer.trackedFrequency.floatValue / frequency))
            var noteName = String(format: "%@%d", noteNamesWithSharps[index], octave)
            noteNameWithSharpsLabel.stringValue = noteName
            noteName = String(format: "%@%d", noteNamesWithFlats[index], octave)
            noteNameWithFlatsLabel.stringValue = noteName
        }
        amplitudeLabel.stringValue = String(format: "%0.2f", analyzer.trackedAmplitude.floatValue)
    }
}
